core/app/models/spree/line_item.rb

```diff
@@ -138,49 +114,68 @@ def options=(options={})
       opts = options.dup # we will be deleting from the hash, so leave the caller's copy intact
 
       currency = opts.delete(:currency) || order.try(:currency)
+      set_price(currency, opts)
+
+      self.assign_attributes opts
+    end
 
+    private
+    # Sets the quantity to zero if it is nil or less than zero.
+    def invalid_quantity_check
+      self.quantity = 0 if quantity.nil? || quantity < 0
+    end
+
+    # Sets this line item's  tax category, price, cost price, and currency from
+    # its variant if they are nil and a variant is present.
+    def copy_variant_attributes
+      return unless variant
+      self.tax_category ||= variant.tax_category
+      self.cost_price ||= variant.cost_price
+      self.currency ||= variant.currency
+      set_price
```

Comment:
I wouldn't call this method implicitly. I would add this to the `before_validation` call, because more explicit.

Also, you can remove the redundant `self.` here. Although not activated, make the hound happy :)